File: business/src/main/java/com/kevinguanchedarias/owgejava/business/unit/ObtainedUnitEventEmitter.java
package com.kevinguanchedarias.owgejava.business.unit;

import com.kevinguanchedarias.owgejava.business.AsyncRunnerBo;
import com.kevinguanchedarias.owgejava.business.SocketIoService;
import com.kevinguanchedarias.owgejava.business.UnitTypeBo;
import com.kevinguanchedarias.owgejava.business.user.UserEventEmitterBo;
import com.kevinguanchedarias.owgejava.business.util.TransactionUtilService;
import com.kevinguanchedarias.owgejava.entity.ObtainedUnit;
import com.kevinguanchedarias.owgejava.entity.UserStorage;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@AllArgsConstructor
public class ObtainedUnitEventEmitter {
    public static final String UNIT_OBTAINED_CHANGE = "unit_obtained_change";
    public static final String UNIT_TYPE_CHANGE = "unit_type_change";

    private final TransactionUtilService transactionUtilService;
    private final SocketIoService socketIoService;
    private final ObtainedUnitFinderBo obtainedUnitFinderBo;
    private final AsyncRunnerBo asyncRunnerBo;
    private final UnitTypeBo unitTypeBo;
    private final UserEventEmitterBo userEventEmitterBo;

    public void emitObtainedUnitsAfterCommit(UserStorage user) {
        transactionUtilService.doAfterCommit(() -> emitObtainedUnits(user));
    }

    public void emitObtainedUnits(UserStorage user) {
        socketIoService.sendMessage(user, UNIT_OBTAINED_CHANGE,
                () -> obtainedUnitFinderBo.findCompletedAsDto(user));
    }

    /**
     * Emits changes to elements affected by a unit alteration, for example, energy, or type's count
     */
    public void emitSideChanges(List<ObtainedUnit> obtainedUnits) {
        if (!obtainedUnits.isEmpty()) {
            var user = obtainedUnits.get(0).getUser();
            var userId = user.getId();

            if (isOneUnitHavingEnergy(obtainedUnits)) {
                asyncRunnerBo.runAssyncWithoutContextDelayed(() -> userEventEmitterBo.emitUserData(user));
            }
            asyncRunnerBo.runAssyncWithoutContextDelayed(() ->
                    socketIoService.sendMessage(userId, UNIT_TYPE_CHANGE, () -> unitTypeBo.findUnitTypesWithUserInfo(userId))
            );
            emitObtainedUnits(user);
        }
    }

    private static boolean isOneUnitHavingEnergy(List<ObtainedUnit> obtainedUnits) {
        return obtainedUnits.stream()
                .map(obtainedUnit -> obtainedUnit.getUnit().getEnergy())
                .anyMatch(energy -> energy != null && energy > 0);
    }
}

File: rust-backend/owge-business/src/bo/unlocked/obtained_unit_event_emitter.rs
//! Port of `ObtainedUnitEventEmitter`
//! (`com.kevinguanchedarias.owgejava.business.unit.ObtainedUnitEventEmitter`).
//!
//! Three emitters:
//! * `emit_obtained_units`  — `unit_obtained_change`
//! * `emit_unit_type_change` — `unit_type_change`
//! * `emit_side_changes`    — composite: optionally `user_data_change`, then
//!                            `unit_type_change` + `unit_obtained_change`.
//!
//! The caller passes `any_unit_has_energy: bool` to `emit_side_changes`, mirroring
//! Java's `isOneUnitHavingEnergy` predicate (unit.energy != null && energy > 0).
//! Computing that predicate is a call-site concern; the emitter only acts on it.

use crate::bo::emitter::unit_type_emitter::UnitTypeEmitter;
use crate::error::OwgeResult;
use crate::websocket::emitter;
use sqlx::MySqlConnection;

pub struct ObtainedUnitEventEmitter;

impl ObtainedUnitEventEmitter {
    /// Emits `unit_obtained_change` — the list of completed obtained-unit DTOs.
    ///
    /// Java: `ObtainedUnitEventEmitter.emitObtainedUnits` ->
    /// `socketIoService.sendMessage(user, UNIT_OBTAINED_CHANGE, () -> obtainedUnitFinderBo.findCompletedAsDto(user))`.
    pub async fn emit_obtained_units(conn: &mut MySqlConnection, user_id: i32) -> OwgeResult<()> {
        emitter::send_message(conn, user_id, "unit_obtained_change", |conn| {
            Box::pin(async move {
                Ok(serde_json::to_value(
                    crate::bo::ObtainedUnitBo::find_completed_dtos(&mut *conn, user_id).await?,
                )?)
            })
        })
        .await
    }

    /// Emits the bundle of side-effects caused by a unit alteration.
    ///
    /// Java: `ObtainedUnitEventEmitter.emitSideChanges(List<ObtainedUnit>)`.
    /// * If `any_unit_has_energy` — emit `user_data_change` (energy totals changed).
    /// * Always emit `unit_type_change` + `unit_obtained_change`.
    ///
    /// The caller is responsible for determining `any_unit_has_energy`
    /// (i.e. whether any affected unit has `energy > 0`).
    pub async fn emit_side_changes(
        conn: &mut MySqlConnection,
        user_id: i32,
        any_unit_has_energy: bool,
    ) -> OwgeResult<()> {
        if any_unit_has_energy {
            crate::bo::user_event_emitter::UserEventEmitter::emit_user_data(conn, user_id).await?;
        }
        UnitTypeEmitter::emit_unit_type_change(conn, user_id).await?;
        Self::emit_obtained_units(conn, user_id).await?;
        Ok(())
    }
}
